redash/query_runner/json_ds.py

```diff
@@ -164,32 +167,80 @@ def run_query(self, query, user):
 
         fields = query.get("fields")
         path = query.get("path")
+        pagination = query.get("pagination", {"type": None})
 
         if isinstance(request_options.get("auth", None), list):
             request_options["auth"] = tuple(request_options["auth"])
         elif self.configuration.get("username") or self.configuration.get("password"):
-            request_options["auth"] = (
-                self.configuration.get("username"),
-                self.configuration.get("password"),
-            )
+            request_options["auth"] = (self.configuration.get("username"), self.configuration.get("password"))
 
         if method not in ("get", "post"):
             raise QueryParseError("Only GET or POST methods are allowed.")
 
         if fields and not isinstance(fields, list):
             raise QueryParseError("'fields' needs to be a list.")
 
-        response, error = self.get_response(query["url"], http_method=method, **request_options)
+        if not isinstance(pagination, dict) or not isinstance(pagination.get("type"), (str, type(None))):
+            raise QueryParseError("'pagination' should be an object with a `type` property")
 
+        results, error = self._get_all_results(query["url"], method, path, pagination, **request_options)
         if error is not None:
             return None, error
 
-        data = json_dumps(parse_json(response.json(), path, fields))
-
+        data = json_dumps(parse_json(results, fields))
         if data:
             return data, None
-        else:
-            return None, "Got empty response from '{}'.".format(query["url"])
+        return None, "Got empty response from '{}'.".format(query["url"])
+
+    def _get_all_results(self, url, method, result_path, pagination, **request_options):
+        """Get all results from a paginated endpoint."""
+        base_url = self.configuration.get("base_url")
+        url = urljoin(base_url, url)
+
+        results = []
+        has_more = True
+        while has_more:
+            response, error = self._get_json_response(url, method, **request_options)
+            has_more = False
+
+            result = _normalize_json(response, result_path)
+            if result:
```

Comment:
Having this many nested `if` statements is kind of a code smell. Maybe we should split this up into multiple functions?

Reply:
In my opinion, yes - definitely worth it if it makes the code easier to follow.